### PR TITLE
feat: support user authentication in frontend

### PR DESCRIPTION
**Pull request by yezizp2012**
## What's changed and what's your intention?

As title, until now we can create a new user to login, the frontend will do the authentication check if password provided. Currently we using ~~`risingwave`~~ `root` as default super user and all e2e test is running under ~~`risingwave`~~ `root`. 
image

Note that: 
`psql` use `-U, --username=USERNAME` as user option, `sqllogictest` uses `-u`/`--user`.

Current user is already recorded in each session, this is helpful for #2814. Cc @cnissnzg 

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Review by BowenXiao1999 — COMMENTED** [review]
1. Maybe we should unify the `-u` and `-U` later ..
2. So for now, we can log in as `risingwave` without password. But we need password for any other users?
3. Do we set different permissions for different users? For example, superuser and other default user.

Rest LGTM

**Comment by yezizp2012** [comment]
> 1. Maybe we should unify the `-u` and `-U` later ..
> 
> 2. So for now, we can log in as `risingwave` without password. But we need password for any other users?
> 
> 3. Do we set different permissions for different users? For example, superuser and other default user.
> 
> 
> 
> Rest LGTM

NTFS for 1. `risingwave` is the default initialized super user in our cluster. We can create any user with different options like can_login/superuser etc, and also with password or not. If provided, the new user will be required to enter password when log in. You can find more details in #2643. The permission check for privilege is not implemented yet.

File: src/frontend/src/session.rs [hunk 1]
```diff
@@ -359,6 +372,8 @@ impl SessionImpl {
         Self {
             env: FrontendEnv::mock(),
             database: "dev".to_string(),
+            user_name: "risingwave".to_string(),
```
Comment (fuyufjh) [inline src/frontend/src/session.rs:375]: May use the global constants e.g. `DEFAULT_DATABASE_NAME` and `DEFAULT_SUPPER_USER`

File: src/frontend/src/session.rs [hunk 2]
```diff
@@ -420,7 +443,53 @@ impl SessionManager for SessionManagerImpl {
...
+                self.env.clone(),
+                database.to_string(),
+                user_name.to_string(),
+                authenticator,
+            )
+            .into())
+        } else {
+            Err(Box::new(Error::new(
+                ErrorKind::InvalidInput,
+                format!("Not found user name: {}", user_name),
```
Comment (fuyufjh) [inline src/frontend/src/session.rs:490]:
```suggestion
                format!("role {} does not exist", user_name),
```

File: README.md [hunk 3]
```diff
@@ -57,7 +57,7 @@ To connect to the RisingWave server, you will need to [install PostgreSQL shell]
 
 ```shell
 # Use psql to connect RisingWave cluster
-psql -h localhost -p 4566 -d dev
+psql -h localhost -p 4566 -d dev -U risingwave
```
Comment (fuyufjh) [inline README.md:60]: Why not "root"? It's more intuitive to me, and shorter ☺️

**Review by fuyufjh — APPROVED** [review]
LGTM. Please reconsider the name of root user :lark-cry:

File: README.md (hunk 3 above)
Comment (TennyZhuang) [reply README.md:60]: Whatever, and materialize uses materialize (

Comment (yezizp2012) [reply README.md]: Both are ok. Well, shorter will be more convenient. I will use root instead. 😊

**Comment by hengm3467** [comment]
@yezizp2012  Is it accurate to say that RisingWave now supports "CREATE USER"? Is the syntax same with PG's? Thanks!

**Comment by yezizp2012** [comment]
> @yezizp2012 Is it accurate to say that RisingWave now supports "CREATE USER"? Is the syntax same with PG's? Thanks!

Yes, RisingWave supports "CREATE/DROP USER" now. The syntax is quit the same with PG, but only support some options as follows (some options in PG are not supported currently):
```
CREATE USER name [ [ WITH ] option [ ... ] ]

where option can be:

      SUPERUSER | NOSUPERUSER
    | CREATEDB | NOCREATEDB
    | LOGIN | NOLOGIN
    | [ ENCRYPTED ] PASSWORD 'password' | PASSWORD NULL

DROP USER [ IF EXISTS ] name [, ...]
```
`ALTER USER` is gonna be supported in the future.

**Comment by CharlieSYH** [comment]
@yezizp2012 
I tested this statement and it retuned an error "ERROR:  sql parser error: Expected end of statement, found: PASSWORD":
`CREATE USER u1 PASSWORD 's1234';`
But this works: `CREATE USER u1 WITH PASSWORD 's1234'`

Seems that 'with' is not optional. Can you confirm on that?

**Comment by yezizp2012** [comment]
> @yezizp2012 
> 
> I tested this statement and it retuned an error "ERROR:  sql parser error: Expected end of statement, found: PASSWORD":
> 
> `CREATE USER u1 PASSWORD 's1234';`
> 
> But this works: `CREATE USER u1 WITH PASSWORD 's1234';`
> 
> 
> 
> Seems that 'with' is not optional. Can you confirm on that?

Yes, 'with' is required.

**Comment by yezizp2012** [comment]
Will be fix in #4414 . @CharlieSYH

**Comment by CharlieSYH** [comment]
Can you point out which one in each option pair is the default if not specified? @yezizp2012 

      SUPERUSER | NOSUPERUSER
      CREATEDB | NOCREATEDB
      LOGIN | NOLOGIN
      [ ENCRYPTED ] PASSWORD 'password' | PASSWORD NULL

**Comment by yezizp2012** [comment]
> Can you point out which one in each option pair is the default if not specified? @yezizp2012
> 
> ```
>   SUPERUSER | NOSUPERUSER
>   CREATEDB | NOCREATEDB
>   LOGIN | NOLOGIN
>   [ ENCRYPTED ] PASSWORD 'password' | PASSWORD NULL
> ```

The default options for all are: `NOSUPERUSER`, `NOCREATEDB`, `LOGIN`, `PASSWORD NULL`.

**Comment by CharlieSYH** [comment]
> > Can you point out which one in each option pair is the default if not specified? @yezizp2012
> > ```
> >   SUPERUSER | NOSUPERUSER
> >   CREATEDB | NOCREATEDB
> >   LOGIN | NOLOGIN
> >   [ ENCRYPTED ] PASSWORD 'password' | PASSWORD NULL
> > ```
> 
> The default options for all are: `NOSUPERUSER`, `NOCREATEDB`, `LOGIN`, `PASSWORD NULL`.

That's awkward. Then I can create a database with a user role that is not supposed to have CREATEDB previledge:

```
dev=> create user u8;
CREATE_USER
dev=> \q
shenyuhan@shenyuhans-MacBook-Pro risingwave % psql -h localhost -p 4566 -d dev -U u8;
psql (14.4, server 9.5.0)
Type "help" for help.

dev=> create database db1;
CREATE_DATABASE
```

**Pull request merged**